File: leetcode/medium/95-unique-bst-2/rec-dp.py
"""
  @author Vortexx2
  Problem 95 - Unique BST 2
  @approach This approach uses a dictionary as a DP approach to be able to reduce the memory used in the recursive stack.

  Runtime - 102 ms
  Memory Usage - 14.9 MB
"""
from typing import *


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
    cache = {}

    def helper(start: int, end: int) -> List[Optional[TreeNode]]:
      if start > end:
        return [None]

      if (start, end) in cache: return cache[(start, end)]

      res = []
      for i in range(start, end + 1):
        left = helper(start, i - 1)
        right = helper(i + 1, end)

        for leftNode in left:
          for rightNode in right:
            root = TreeNode(i, leftNode, rightNode)
            res.append(root)

      cache[(start, end)] = res
      return res

    return helper(1, n)

File: leetcode/medium/64-min-path/rec-dp.cpp
/**
 * @file rec-dp.cpp
 * @author Vortexx2
 * @brief Problem 64 - Minimum Path Sum
 * @approach Recursive DP approach
 * @date 03-08-2022
 * 
 * Runtime - 8ms
 * Memory Usage - 10.2 MB
 */
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<vector<int>> dp;
  int minPathSum(vector<vector<int>>& grid) {
    dp.resize(grid.size(), vector<int>(grid[0].size(), -1));
    return helper(grid, 0, 0);
  }

  int helper(vector<vector<int>>& grid, int row, int col) {
    if (row == grid.size() - 1 && col == grid[0].size() - 1)
      return grid[row][col];

    if (dp[row][col] != -1) return dp[row][col];

    if (row == grid.size() - 1)
      dp[row][col] = grid[row][col] + helper(grid, row, col + 1);

    else if (col == grid[0].size() - 1)
      dp[row][col] = grid[row][col] + helper(grid, row + 1, col);

    else
      dp[row][col] = grid[row][col] + min(helper(grid, row + 1, col),
                                          helper(grid, row, col + 1));

    return dp[row][col];
  }
};